### ISSUE TITLE
Enable hybrid processors support with oneTBB (tbbbind library)

### ISSUE THREAD

**Issue by moslex** [Homebrew/homebrew-core#109772]
### Provide a detailed description of the proposed feature

Hello, Brew community.

I’d like to make request to add the latest version of oneTBB with hybrid processors support into brew distribution channel.
Current version of oneTBB in the brew distribution does not support hybrid processors, such as M1, M2, Arm or [Alder Lake](https://ark.intel.com/content/www/us/en/ark/products/codename/147470/products-formerly-alder-lake.html). 
To enable this support, please follow these instructions: https://github.com/oneapi-src/oneTBB/blob/master/cmake/README.md
Important things:
•	The oneAPI Threading Building Blocks (oneTBB) implementation requires [the hwloc library](https://www-lb.open-mpi.org/projects/hwloc) to query NUMA(version >= 1.11) and Hybrid CPUs (version >= 2.4) topology information. Brew: https://formulae.brew.sh/formula/hwloc; HWLOC project: https://www-lb.open-mpi.org/projects/hwloc/ 
•	TBBBind library configuration:
o	tbbbind links with HWLOC 1.11.x
o	tbbbind_2_0 links with HWLOC 2.1–2.4
o	tbbbind_2_5 links with HWLOC 2.5 and later
•	Automatic HWLOC searching requires CMake version 3.6 or higher.
•	HWLOC automatic search is managed through pkg-config: https://formulae.brew.sh/formula/pkg-config

### What is the motivation for the feature?

More effective threading -> better performance -> less power is required -> save the Planet!
Let’s give a chance to modern and fancy hybrid processors! 😊 

### How will the feature be relevant to at least 90% of Homebrew users?

Hybrid processors are popular now.
So, this feature will be more and more valid for all users with new hybrid HW.

### What alternatives to the feature have been considered?

Boring single thread execution or less effective threading mechanism which does not manage high\low power cores.

**Comment by SMillerDev** [comment]
Feature requests should be made as pull requests

**Comment by carlocab** [comment]
Do you have benchmarks on the performance improvement from building TBBBind? What about benchmarks from doing the same when TBB is built using `brew`? Does it require specific optimisation flags (e.g. `-march`, `-mcpu`, etc) to build?

**Comment by ilya-lavrenov** [comment]
> Do you have benchmarks on the performance improvement from building TBBBind? What about benchmarks from doing the same when TBB/TBBBind is built using brew?

The TBBBind is about allowing users to distinguish performance / efficient cores on Apple M1 / M2 processor and schedule heavy workloads on performance cores. Without having TBBBind users (e.g. https://github.com/openvinotoolkit/openvino) is not able to understand the difference between cores and a workload become to use all cores, while efficient ones need to be kept for users tasks. 

> Does it require specific optimisation flags (e.g. -march, -mcpu, etc) to build?

No, just a usual library. TBB uses this library as a plugin

**Comment by carlocab** [comment]
> The TBBBind is about allowing users to distinguish performance / efficient cores on Apple M1 / M2 processor and schedule heavy workloads on performance cores.

That sounds nice, but also rather niche.

> Without having TBBBind users (e.g. https://github.com/openvinotoolkit/openvino) is not able to understand the difference between cores and a workload become to use all cores, while efficient ones need to be kept for users tasks.

Would the OS manage the workload allocation in this case? The OS can probably get this right in most common use-cases. (I'd probably also argue that the OS is likely better than the developer at deciding this allocation for all but the most experienced at performance-tuning.)

Does TBBBind even work on macOS? Their [binary release](https://github.com/oneapi-src/oneTBB/releases/tag/v2021.5.0) does not include it on macOS, but does on Linux.

**Comment by ilya-lavrenov** [comment]
> Wouldn't the OS manage the workload allocation in this case? The OS can probably get this right in most common use-cases

For simple workloads yes, but when we want to understand the maximum capability of the system (e.g. how many FPS it can process), we need to estimate this value based on the resources and here it's important to distinguish cores. 
Example: if you have 2e + 8p cores, you can execute 8 abstract tasks in parallel with the same latency. If you would think that you have 10 equal cores, latency will change from frame to frame depending on which cores is used to execute the task

I would suggest to put this discussion about the objective on hold - I suppose TBB developers know better why they created tbbbind library.. Our goal is to discuss the packaging aspects

> Does TBBBind even work and/or build on macOS? Their [binary release](https://github.com/oneapi-src/oneTBB/releases/tag/v2021.5.0) does not include it on macOS, but does on Linux.

Prebuilt packages are provided only for x86 and I would suppose there is not need to distinguish cores on x86 Apple hosts
When we are talking about ARM - this is definitely needed for our use cases.
TBBBind is based on hwloc, so if hwloc is available on Apple M1, than TBBBind should also work since it's just a thin layer solving hwloc compatibility issues

**Comment by carlocab** [comment]
> I would suggest to stop this discussion about the purpose of TBB components - I suppose TBB developers know better why they created tbbbind library.. Our goal is to discuss the packaging aspects

I'm afraid you've misunderstood my point. I was discussing packaging aspects, and arguably the most important one: usefulness. If we're not even sure that enabling this is useful, then it's unlikely we'd be interested in enabling this feature in the formula. If it useful, then we can consider it.

**Comment by ilya-lavrenov** [comment]
> usefulness

I've provided the scenarios why TBB users needs information about system topology which can be taken via TBBBind with hwloc underneath. Yes, it's not typical when you just want to split a specific work by threads, but more specific and concrete:
- if user wants **stable performance with max latency** - the work is definitely must be schedule on big cores. In this scenario OS does not guarantee that your workload will not be moved from core to core, that is why TBB provides so-called constraints (on implementation level, it's mapped on TBB's `core_type_id` in `constraints` structure https://oneapi-src.github.io/oneTBB/main/reference/constraints_extensions.html) which allows to create a specific group of threads (TBB arena), which will be scheduled only on big cores. Without having TBBBind it's not possible on Apple M1
- if user wants max FPSs number, the work can be scheduled on all available types of cores.

Usual users who just browsing the internet can rely on the OS as you wrote.

> I'm afraid you've misunderstood my point

I well understood your point, thanks :) but my point is that people cannot be experts in all fields unfortunately at the same time. That is why we can rely on library developers (taking into account that TBB is well-known library with many years history), if library developers created a new library / component, they know why they did it, is on it? TBB is Threading Building Blocks library and allows multiple advanced levels of control on how users tasks are executed in threads arena, why should we limit TBB functionality? ;)

This specific component is definitely applicable for ARM since it's hybrid architecture.

**Comment by SMillerDev** [comment]
> I would consider this as a bug, because important functionality is not exposed

It's a new feature you want enabled, it's a feature request.

**Comment by carlocab** [comment]
> * if user wants **stable performance with min latency** (multiple applications in AI like face recognition) -  - the work is definitely must be schedule on big cores. In this scenario OS does not guarantee that your workload will not be moved from core to core, that is why TBB provides so-called constraints (on implementation level, answer from TBBBind 2.5 is mapped on TBB's `core_type_id` in `constraints` structure https://oneapi-src.github.io/oneTBB/main/reference/constraints_extensions.html) which allows to create a specific group of threads (TBB arena), which will be scheduled only on big cores. Without having TBBBind it's not possible on Apple silicon
> * if user wants max FPSs number, the work can be scheduled on all available types of cores.

Do we ship any formula that would make use of this?

> why should we limit TBB functionality? ;)

Because adding functionality has the cost of an extra dependency. Enabling this functionality means that: all users who `brew install tbb`, or `brew install` something that depends on TBB must now also install `hwloc` even if they will not use the features it adds. Adding the dependency also increases the workload that needs to be done by our CI infrastructure, and we shouldn't be adding to that unless there is a very good reason for it.

**Comment by ilya-lavrenov** [comment]
> Because adding functionality has the cost of an extra dependency. Enabling this functionality means that: all users who brew install tbb, or brew install something that depends on TBB must now also install hwloc even if they will not use the features it adds

It can organized this way: since TBB loads TBBBind as a plugin, so we can have several packages - TBB and TBBBind. 
Users who works with hybrids cores can additionally install one package.

> Do we ship any formula that would make use of this?

I don't know. I'm just a TBB and brew user and want to install software using package managers instead of compiling it by myself and redistributing. Once we have such functionality, we can bring our own package that depends on TBBBind to brew, but plans are not confirmed yet.

**Comment by ilya-lavrenov** [comment]
Fixed here https://github.com/Homebrew/homebrew-core/pull/112805

**Comment by carlocab** [comment]
Reverted in ea7c5a51ab9fc0321aa30b79143c01b642c86faa.

TBBBind is not supported on macOS. See #112805 for details.
